Task 77372

AlexandrSid opened the pull request:
https://faang-school.atlassian.net/browse/BJS2-77372
Постановка целей
Task done
Надо бы добавить тесты на маппер и Post/Get/Update/DeleteMapping на контроллер, в качестве дополнения

AlexandrSid committed:
Task 77372
Постановка целей
Создать новую цель - done

AlexandrSid committed:
Task 77372
Постановка целей
Обновить цель - done

AlexandrSid committed:
Task 77372
Постановка целей
Обновить цель - completely done

AlexandrSid committed:
Task 77372
Постановка целей
Удалить цель - done

AlexandrSid committed:
Task 77372
Постановка целей
Done

AlexandrSid committed:
Task 77372
Постановка целей
Builder Fix

littlerock2877 commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,46 @@
+    public GoalDto createGoal(Long userId, Goal goal) {
+        if (!goal.getTitle().isBlank()) throw new IllegalArgumentException("Goal has no title");
+        Goal createdGoal = goalService.createGoal(userId, goal);
+        return goalMapper.goalToGoalDTO(createdGoal);
```
А оно точно будет работать с аннотацией @Controller? ты пытаешься вернуть ДТО, для этого есть аннотация @ResponseBody, которая включена в @RestController. Мб стоит заменить?

AlexandrSid replied:
РЕСТа на данном этапе не добавлено

littlerock2877 replied:
Но у тебя же запросы просто не работают) По сути у тебя сейчас написан код ради кода просто из-за одной маленькой детали)

littlerock2877 replied:
Я не досмотрел лекции, и думал, что нужно было полностью реализовать на этом спринте. Тут я не прав)

littlerock2877 commented on src/main/java/school/faang/user_service/dto/goal/GoalFilterDto.java:
```diff
@@ -0,0 +1,33 @@
+    private String title;
+    private GoalStatus status;
+    private List<String> skillTitles;
+    //todo goal_created_before and after
```
Все равно скажут убрать коммент на ревью, вдруг ты не заметил)

AlexandrSid replied:
может этот функционал будет нужен, если нужен, то в каком формате стоит передавать дату не продумано ещё, станет ясно - можно или реализовать или стререть

littlerock2877 replied:
Вроде бы вы с Михаилом обсуждали этот момент на первом этапе, и сошлись на том, что на групповом проекте комментарии в коде не нужны, прям совсем) Сейчас это звучит как "может надо а может нет", но если вдруг кто-то другой сядет доделывать эту задачу - что ему скажет этот комментарий?)

littlerock2877 commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,174 @@
+
+    private final GoalMapper goalMapper;
+
+    private final GoalRepository goalRepository;
```
а зачем разделять между собой поля с одинаковым модификатором доступа?) На 1 модуле придирались к этому

AlexandrSid replied:
у меня не придирались)
Классы разные, не объединять же их в одну кучу только по признаку доступности

littlerock2877 replied:
Здесь согласен, это скорее придирка) Мне в первом модуле в каждом мре писали за эти пустые строчки между полями класса)

littlerock2877 commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,174 @@
+                .count();
+
+        if (usersActiveGoals >= MAXIMUM_ALLOWED_ACTIVE_GOALS) {
+            throw new IllegalArgumentException("User exceeded maximum allowed number or active goals "
```
Лучше использовать String.format или "".formatted для конкатенации строк

AlexandrSid replied:
с исключениями вообще стоит продумать, что в них писать пока. Текст в некоторых излишен - по типу понятно было бы, что не сработало. Хотя конкретно здесь без сообщения не обойтись

littlerock2877 replied:
Здесь не совсем про это идет речь) Я имеб ввиду, что сейчас при конкатенации ты создаешь сначала первую String переменную, потом когда плюсуешь - то создается уже вторая. Лучше использовать для этого String.format.

littlerock2877 commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,174 @@
+        skills.forEach(skill -> skill.getGoals().remove(goalToDelete));
+        skillService.updateAll(skills);
+
+        //var invitations = goalToDelete.getInvitations();
```
Комментарий

AlexandrSid replied:
напрямую относитстя к следующей задаче - про приглашения к совместной работе над целями

littlerock2877 replied:
Ну, как я и писал выше, имхо но комменты вообще не нужны) А в этом плане еще добавляется то, что это совсем не относится к текущей задаче) Это то же самое, что если бы ты залил ПР с какой-то фичей, но во время выполнения заметил пару багов из мастера, сразу не создавая отдельный ПР пофиксил бы все эти баги. Как мне кажется, задачи все таки нужно разделять

AlexandrSid replied:
вообще это туду, а не просто коммент. Мне же над этим рабоать, может сегодня, а может через неделю, а может этого вообще не придётся делать, а может это будет кто-то ещё делать (в таком случае маловато данных, конечно).
 Мы вроде комменты договорились не оставлять пототму, что код должен быть self-explained, а не потому, что это табу. Как и зачем с помощью кода помечать места, которые нужно проработать вне рамок выполняемой задачи? 

В конеченом итоге этого туду не должно остаться, разумеется. А пока это рабочий момент.

littlerock2877 replied:
Если вы реально на работе заливаете закомменченные куски кода в мастер чтобы потом доработать, то сочувствую) Но ладно, мне лень спорить)

JegorKuz commented on src/main/java/school/faang/user_service/dto/goal/GoalFilterDto.java:
```diff
@@ -0,0 +1,33 @@
+    private List<String> skillTitles;
+    //todo goal_created_before and after
+
+    public boolean doFilter(Goal goal) {
```
Разве DTO не подразумевает что этот класс содержит исключительно данные и никакой логики?

AlexandrSid replied:
лень было маппер писать ещё и на него. Вообще там не дто должен быть, а GoalFilter, и дто для него пока выглядит избыточно, но по требованиям задачи - дто

JegorKuz replied:
Но ведь у тебя есть класс GoalFilterDto? И вопрос был скорее про метод в дто который выполняет какой-то функционал, правильно понимаю что это больше как временное решение пока?

JegorKuz commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,174 @@
+    public Goal updateGoal(Long goalId, GoalDto goalDto) {
+        var goalToUpdate = goalRepository.findById(goalId)
+                .orElseThrow(() -> new NoSuchElementException("No goal found by id " + goalId));
+        goalMapper.updateGoalFromDto(goalDto, goalToUpdate);
```
В условии "Обновить цель" 5ый пункт ведь сказано что сначала должны быть проверки на:
1) Нужно проверить: если пользователь завершает цель (поле status), то ее можно обновить только если она не была завершена ранее. (Эту проверку не вижу)
2) Также проверить, что цель содержит только существующие скиллы (Эту тоже не вижу)

JegorKuz commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,174 @@
+                .toList();
+    }
+
+    public List<Goal> getGoalsByUser(Long userId, GoalFilterDto filter) {
```
В задании сказано что метод должен называться "findGoalsByUserId" что совпадает с логикой ибо ищем не по юзеру а по его id.

JegorKuz commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,174 @@
+        //todo on next task with invitations
+    }
+
+    public List<Goal> findSubtasksByGoalId(long goalId) {
```
А зачем фильтровать по пустому фильтру? разве не проще тогда вернуть все возможные результаты без фильтров, что по сути тоже самое?

JegorKuz commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,174 @@
+        return goalToDelete;
+    }
+
+    private void deleteGoalRecursive(Goal goalToDelete) {
```
Почему рекурсивно если рекурсия не используется?

JegorKuz reviewed (commented):
Оставил коменты касательно оформления и немного  не понял по логике местами.

AlexandrSid committed:
Task 77372
Постановка целей
Refactored after CrossReview

AlexandrSid committed:
Task 77372
Постановка целей
Сервисы реализованы через интерфейсы

AlexandrSid committed:
Task 77372
Постановка целей
Фильтр реализован через стратегии

AlexandrSid committed:
Task 77372
Постановка целей
Minor refactoring

miked4king commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,46 @@
+    private final GoalMapper goalMapper;
+
+    public GoalDto createGoal(Long userId, Goal goal) {
+        if (!goal.getTitle().isBlank()) throw new IllegalArgumentException("Goal has no title");
```
точно правильная проверка?
исключение выбрасывается, если title не пустой

miked4king commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public GoalDto updateGoal(Long goalId, GoalDto goalDto) {
+        if (!goalDto.getTitle().isBlank()) throw new IllegalArgumentException("Goal has no title");
```
точно правильная проверка?

miked4king commented on src/main/java/school/faang/user_service/dto/goal/GoalDto.java:
```diff
@@ -0,0 +1,21 @@
+@Data
+public class GoalDto {
+    private Long id;
+
```
давай уберем лишние отступы между полями

miked4king commented on src/main/java/school/faang/user_service/mapper/GoalMapper.java:
```diff
@@ -0,0 +1,48 @@
+
+    @Mapping(source = "parent.id", target = "parentId")
+    @Mapping(target = "skillIds", expression = "java(goalSkillsToSkillIDs(goal))")
+    GoalDto goalToGoalDTO(Goal goal);
```
можно улучшить названия методов для мапппера, как правило называют toDto, toEntity.
например для этого сделать toGoalDto.
давай над другими тоже подумаем

miked4king commented on src/main/java/school/faang/user_service/mapper/GoalMapper.java:
```diff
@@ -0,0 +1,48 @@
+
+    List<GoalDto> mapGoalsToDTOs(List<Goal> goals);
+
+    List<Goal> mapGoalsDTOsToGoals(List<GoalDto> dtos);
```
используется где нибудь?

miked4king commented on src/main/java/school/faang/user_service/mapper/GoalMapper.java:
```diff
@@ -0,0 +1,48 @@
+    void updateGoalFromDto(GoalDto dto, @MappingTarget Goal goal);
+
+    @Named("mapParent")
+    default Goal mapParent(Long id, @Context GoalService goalService) {
+        return id != null ? goalService.findById(id) : null;
+    }
+
+    @Named("mapSkills")
+    default List<Skill> mapSkills(List<Long> ids, @Context SkillService skillService) {
+        return ids == null ? List.of() : ids.stream()
+                .map(skillService::findById)
+                .toList();
+    }
```
вызов сервиса внутри маппера — не очень хорошая практика, потому что мапперы обычно не должны иметь бизнес-логику или зависеть от сервисов. Это нарушает принцип разделения ответственности.
в маппер обычно передают уже готовые данные

miked4king commented on src/main/java/school/faang/user_service/service/GoalService.java:
```diff
@@ -0,0 +1,33 @@
+
+    Goal findById(Long id);
+
+    public static boolean goalIsActive(@NotNull Goal goal) {
+        return GoalStatus.ACTIVE == goal.getStatus();
+    }
+
+    public static boolean goalIsCompleted(@NotNull Goal goal) {
+        return GoalStatus.COMPLETED == goal.getStatus();
+    }
```
касаемо двух этих методов.
1) обычно используют стиль методов с префиксом is или has для булевых значений, например isGoalActive
2) public можно не указывать явно, т.к это модификатор по умолчанию
3) goalIsCompleted - не нашел где используется, мб убрать?
4) кажется что эти методы не связаны с контрактом интерфейса, а скорее являются утилитарными функциями, думаю лучше разместить либо в утилитном классе, либо вообще не выносить в отдельный метод, т.к используется в одном месте

miked4king commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,190 @@
+    public static int MAXIMUM_ALLOWED_ACTIVE_GOALS = 3;//todo вынести в конфигурацию компонента
+
+    private final GoalMapper goalMapper;
+
```
давай уберем лишние отступы между полями

miked4king commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,190 @@
+
+    private final SkillService skillService;
+    private final UserService userService;
+
```
лишний отступ

miked4king replied:
остался лишний отступ

miked4king commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,190 @@
+                    + usersActiveGoals);
+        }
+
+        var skillsOfUser = skillService.findAllByUserId(userId);
+        var missingSkills = goal.getSkillsToAchieve().stream()
```
вроде договаривались типизировать переменные, давай заменим все var

miked4king commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,190 @@
+    @Override
+    public Goal findById(Long id) {
+        return goalRepository
+                .findById(id)
```
вызов .findById(id) можно не переносить на новую строку

miked4king commented on src/main/java/school/faang/user_service/filter/goal/GoalCreateAfterFilter.java:
```diff
@@ -0,0 +1,12 @@
+
+import java.time.LocalDateTime;
+
+public record GoalCreateAfterFilter(LocalDateTime createAfter) implements GoalFilter {
```
комментарий касаемо всех фильтров.
вижу что реализовал с помощью интерфейса GoalFilter и метода createFilters в утильном классе, который создает список фильтров на основе переданных данных, это хороший подход для объединения нескольких условий фильтрации. 
однако, для более чистой архитектуры и расширяемости рекомендуется:
- зарегистрировать их как Spring бины.
- внедрять список фильтров через DI.

и немного их модифицировать, тоесть у тебя будет метод, который будет определять нужно ли применять фильтр, например  boolean isApplicable(Goal goal);
и будет второй метод, который непосредственно применяет фильтр, как раз логика, которая у тебя в методе doFilter.

у тебя каждый фильтр будет компонентом @Component
в GoalServiceImpl ты можешь заинжектить private List<GoalFilter> filters, спринг автоматически подтянет в этот лист все реализации фильтров. и тебе потребуется для применения фильтров просто проитерироваться по ним.
плюс в том, что если потребуется добавить новый фильтр, тебе придется просто добавить новую реализацию и не изменять код в других местах

miked4king commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public GoalDto deleteGoal(long goalId) {
+        Goal deletedGoal = goalService.deleteGoal(goalId);
```
методы сервиса возвращают entity, что не является лучшей практикой.
обычно рекомендуется возвращать DTO из сервисных методов, чтобы скрывать внутренние детали ORM-сущностей и обеспечить более чистую архитектуру
давай поправим, и в других методах тоже

AlexandrSid committed:
Task 77372
рефакторинг в механизме фильтрации

AlexandrSid committed:
Task 77372
рефакторинг: удаление entity из контроллера,
сервисные методы возвращают дто.
замена var на типы, кроме случаев присвоения вызова конструктора.

AlexandrSid committed:
Task 77372
рефакторинг:
Введён бин маппер-контекста чтобы уйти от проброса в маппер сервисов

Переименованы методы маппера

Goal entity убран из контроллера

Константа на максимальное количество активных целей вынесена в application.yaml

Перенос статических методов Goal.isActive в утил-класс

miked4king commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,205 @@
+public class GoalServiceImpl implements GoalService {
+
+    @Value("${logic.constants.max_active_goals}")
+    public static int MAXIMUM_ALLOWED_ACTIVE_GOALS;
```
точно значение подтягивается сюда? 
никогда не видел чтобы в константу подтягивались значения из энвов, она подразумевает неизменяемость, а раз этот параметр может изменять то не стоит его делать static

miked4king commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,205 @@
+    @Override
+    public GoalDto createGoal(Long userId, Goal goal) {
+        long usersActiveGoals = goalRepository.findGoalsByUserId(userId)
+                .filter(GoalUtil::isGoalActive)
```
а зачем нам выносить проверку в отдельный класс Utils? почему не сделать ее тут на месте?
к тому же этот метод вызывается только в одном месте

miked4king commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,205 @@
+                skillService.findAllByIds(copyOfUpdatesSkillIds).stream()
+                        .map(Skill::getId)
+                        .toList()
+        );
```
в целом ок, но мне кажется более понятным будет сделать так.
1) ищем все скилы skillService.findAllByIds, сохраняем в Set айдишники
2) итерируемся по goalDto.getSkillIds() и смотрим, содержится ли id в Set, сохраняем в List

miked4king commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,205 @@
+            throw new IllegalArgumentException("Skill ids not exists: ".formatted(copyOfUpdatesSkillIds.toArray()));
+
+        goalMapper.updateGoalFromDto(goalDto, goalToUpdate);
+        GoalUtil.updateTime(goalToUpdate, LocalDateTime.now());
```
зачем это выносить в Utils? почему тут не можем сделать goalToUpdate.setUpdatedAt(LocalDateTime.now()) ?

miked4king commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    @Transactional
+    void updateUsersWithSkills(Goal completedGoal) {
```
transactional не будет работать, т.к вызов происходит из этого-же класса. для того, чтобы работала можно создать отдельный слой. в рамках задачи пока-что вас не просят заботиться о транзакциях, так что можем просто убрать)
плюс модификатор доступа у метода потерялся

miked4king commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,205 @@
+        List<User> users = completedGoal.getUsers();
+        List<Skill> skills = completedGoal.getSkillsToAchieve();
+        users.forEach(user -> {
+            var merged = new HashSet<>(skills);
```
в рамках буткемпа мы типизируем переменные, давай заменим var)
поменяй еще в других местах, если они есть

miked4king commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,205 @@
+
+    @Override
+    public List<GoalDto> findSubtasksByGoalId(long goalId) {
+        GoalFilterDto blankFilter = new GoalFilterDto();
```
мы получается всегда будем получать пустой фильтр? зачем он нам тогда нужен?
можем мы в контроллере его будем передавать тогда?

miked4king commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    private List<Goal> filterGoals(Stream<Goal> goalStream, GoalFilterDto filterDto) {
+        goalFilters.forEach(goalFilter -> goalFilter.setCriteria(filterDto));
```
Саша, это не потокобезопасно, в текущей реализации setCriteria() вызывается для каждого фильтра перед применением фильтров.
это рискованно, особенно в многопоточной среде, так как GoalFilterDto — это один объект, который может быть изменен одновременно разными потоками.

давай уберем поле private GoalFilterDto criteria; в каждом фильтре, уберем из интерфейса фильтра метод void setCriteria(GoalFilterDto criteria);
и будем передавать объект Goal в метод isApplicable(Goal goal), а в реализации уже проверять необходимое поле

miked4king commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,205 @@
+
+    @Override
+    public List<GoalDto> findSubtasksByGoalId(long goalId, GoalFilterDto filter) {
+        Stream<Goal> goalsByParent = goalRepository.findByParent(goalId);
```
перед тем, как фильтровать и возвращать список целей, нужно преобразовать сущности в dto

miked4king commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,205 @@
+                .filter(GoalFilter::isApplicable)
+                .toList();
+        return goalStream
+                .filter(goal ->
+                        applicableFilters.stream()
+                                .allMatch(goalFilter -> goalFilter.doFilter(goal))
+                )
+                .toList();
```
давай поправим форматирование

miked4king commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,205 @@
+
+    private List<Goal> filterGoals(Stream<Goal> goalStream, GoalFilterDto filterDto) {
+        goalFilters.forEach(goalFilter -> goalFilter.setCriteria(filterDto));
+        List<GoalFilter> applicableFilters = this.goalFilters.stream()
```
зачем тут this при обращении к фильтрам?

miked4king commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,205 @@
+                .toList();
+    }
+
+    @Override
+    public Goal findById(Long id) {
+        return goalRepository.findById(id)
+                .orElseThrow(NoSuchElementException::new);
+    }
```
вроде нигде не вызывается, давай уберем

miked4king commented on src/main/java/school/faang/user_service/service/skill/SkillServiceImpl.java:
```diff
@@ -0,0 +1,40 @@
+
+@Service
+@RequiredArgsConstructor
+public class SkillServiceImpl implements SkillService {
```
по сути мы создали две прокси над репозиторием, я не понимаю их необходимость.
почему мы не можем заинжектить необходимые репозитории в GoalServiceImpl ?
предложу убрать SkillServiceImpl и UserServiceImpl, и вместо них использовать репозитории

miked4king commented on src/main/java/school/faang/user_service/mapper/goal/GoalMapperContextImpl.java:
```diff
@@ -0,0 +1,29 @@
+
+@Component
+@RequiredArgsConstructor
+public class GoalMapperContextImpl implements GoalMapperContext {
```
я не согласен с данным подходом, тут нарушается принцип единой ответственности, и мы создаем маппер, задача которого преобразовать объект А в объект Б, а маппер у нас зависит от репозитория.
лучше, чтобы бизнес-слой (сервисы) заранее получали необходимые объекты (например, Goal, Skill) и передавали их в маппер.

AlexandrSid committed:
Task 77372
рефакторинг по ПР

из маппера удалено всё, что не используется.
фильтры стейтлесс

AlexandrSid committed:
Task 77372
int injection fix

AlexandrSid committed:
Task 77372
DTOs small fix to prepare them to serialization
